pkg/api/kptfile/v1alpha2/types.go

```diff
@@ -0,0 +1,213 @@
+	// Whether a subpackage is local or remote is determined by whether Upstream is specified.
+	Upstream *Upstream `yaml:"upstream,omitempty"`
+}
+
+// Pipeline declares a pipeline of functions used to mutate or validate resources.
+type Pipeline struct {
+	//  Sources defines the source packages to resolve as input to the pipeline. Possible values:
+	//  a) A slash-separated, OS-agnostic relative package path which may include '.' and '..' e.g. './base', '../foo'
+	//     The source package is resolved recursively.
+	//  b) Resources in this package using '.'. Meta resources such as the Kptfile, Pipeline, and function configs
+	//     are excluded.
+	//  c) Resources in this package AND all resolved subpackages using './*'
+	//
+	// Resultant list of resources are ordered:
+	// - According to the order of sources specified in this array.
+	// - When using './*': Subpackages are resolved in alphanumerical order before package resources.
+	//
+	// When omitted, defaults to './*'.
+	// Sources []string `yaml:"sources,omitempty"`
+
+	// Following fields define the sequence of functions in the pipeline.
+	// Input of the first function is the resolved sources.
+	// Input of the second function is the output of the first function, and so on.
+	// Order of operation: mutators, validators
+
+	// Mutators defines a list of of KRM functions that mutate resources.
+	Mutators []Function `yaml:"mutators,omitempty"`
+
+	// Validators defines a list of KRM functions that validate resources.
+	// Validators are not permitted to mutate resources.
+	Validators []Function `yaml:"validators,omitempty"`
+}
+
+// Function specifies a KRM function.
+type Function struct {
+	// `Image` specifies the function container image.
+	// It can either be fully qualified, e.g.:
+	//
+	//	image: gcr.io/kpt-fn/set-label
+	//
+	// Optionally, kpt can be configured to use a image
+	// registry host-path that will be used to resolve the image path in case
+	// the image path is missing (Defaults to gcr.io/kpt-fn).
+	// e.g. The following resolves to gcr.io/kpt-fn/set-label:
+	//
+	//	image: set-label
+	Image string `yaml:"image,omitempty"`
+
+	// `Config` specifies an inline k8s resource used as the function config.
+	// Config, ConfigPath, and ConfigMap fields are mutually exclusive.
+	Config *yaml.Node `yaml:"config,omitempty"`
+
+	// `ConfigPath` specifies a slash-delimited relative path to a file in the current directory
+	// containing a K8S resource used as the function config. This resource is
+	// excluded when resolving 'sources', and as a result cannot be operated on
+	// by the pipeline.
+	ConfigPath string `yaml:"configPath,omitempty"`
+
+	// `ConfigMap` is a convenient way to specify a function config of kind ConfigMap.
+	ConfigMap map[string]string `yaml:"configMap,omitempty"`
+}
+
+// Inventory encapsulates the parameters for the inventory resource applied to a cluster.
+// All of the the parameters are required if any are set.
```

Comment:
I'll follow up with @seans3 to expand the comments here.

Reply:
@martinmaly You are correct--the first three `Inventory` values (Namespace, Name, and InventoryID) are filled in by `kpt live init`. These three values are needed locally 1) to find the inventory object in the cluster, or 2) to first create the inventory object if it does not yet exist in the cluster. Currently, `kpt live init` must first be run before any `kpt live apply`, and a **package not initialized error** will be thrown if this is not the case.

The namespace is calculated from the namespace of the resources in the package, or the `--namespace` flag during `init`, or a `namespace` set in the `kubeconfig` file. The name is currently `inventory-XXXXX` where XXXXX is a random number to ensure no name collision. We will probably change the name to be more meaningful by starting with the name of the directory. The `InventoryID` is a UUID which is added as a label to the cluster resource because we retrieve inventory objects by label (there used to the possibility of multiple inventory objects, and we also may need that capability for nested packages). I'm pretty sure the other two `Inventory` fields (`Labels` and `Annotations`) are added after `init` during the apply lifecycle.

Reply:
@seans3 Do you mind following up and adding a comment on each field here, and specifying whether it's user-specified or kpt managed? This is a public schema for Kpt, so we want each exported field to be well-documented.